aws-lc-rs/src/key_transport.rs

```diff
@@ -0,0 +1,404 @@
+    ///
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    ///
+    pub fn generate(alg: &'static KemAlgorithm) -> Result<Self, Unspecified> {
+        // match alg {
+        //     KemAlgorithm::KYBER512_R3 =>
+        unsafe {
+            let kyber_key = kem_key_generate(alg.id.nid())?;
+            let mut priv_key_bytes = vec![0u8; alg.secret_key_size()];
+            if 1 != EVP_PKEY_get_raw_private_key(
+                *kyber_key,
+                priv_key_bytes.as_mut_ptr(),
+                &mut alg.secret_key_size(),
+            ) {
+                return Err(Unspecified);
+            }
+            Ok(KemPrivateKey {
+                algorithm: alg,
+                pkey: kyber_key,
+                priv_key: priv_key_bytes.into(),
+            })
+        }
+        // }
+    }
+
+    /// Return the algorithm associated with the given KEM private key.
+    #[must_use]
+    pub fn algorithm(&self) -> &'static KemAlgorithm {
+        self.algorithm
+    }
+
+    /// Computes the KEM public key from the KEM private key
+    ///
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    ///
+    pub fn compute_public_key(&self) -> Result<KemPublicKey, Unspecified> {
+        let mut pubkey_bytes = vec![0u8; self.algorithm.public_key_size()];
+        unsafe {
+            if 1 != EVP_PKEY_get_raw_public_key(
+                *self.pkey,
+                pubkey_bytes.as_mut_ptr(),
+                &mut self.algorithm.public_key_size(),
+            ) {
+                return Err(Unspecified);
+            }
+
+            let pubkey_ctx_copy = LcPtr::new(EVP_PKEY_kem_new_raw_public_key(
+                self.algorithm.id.nid(),
+                pubkey_bytes.as_mut_ptr(),
+                self.algorithm.public_key_size(),
+            ))?;
+
+            Ok(KemPublicKey {
+                algorithm: self.algorithm.clone(),
+                pkey: pubkey_ctx_copy,
+                pub_key: pubkey_bytes.into(),
+            })
+        }
+    }
+
+    /// Performs the decapsulate operation using the current KEM private key on the given ciphertext.
+    ///
+    /// `ciphertext` is the ciphertext generated by the encapsulate operation using the KEM public key
+    /// associated with the current KEM private key.
+    ///
+    /// After the decapsulation is finished, `decapsulate` calls `kdf` with the raw shared secret
+    /// from the operation and then returns what `kdf` returns.
+    ///
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    ///
+    pub fn decapsulate<F, R>(&self, ciphertext: &mut [u8], kdf: F) -> Result<R, Unspecified>
+    where
+        F: FnOnce(&[u8]) -> Result<R, Unspecified>,
+    {
+        unsafe {
+            let ctx = LcPtr::new(EVP_PKEY_CTX_new(*self.pkey, null_mut()))?;
+            let mut shared_secret: Vec<u8> = vec![0u8; self.algorithm.shared_secret_size()];
```

Comment:
Does this still need to be allocated on the heap, now that we know the size beforehand? Same question for all the other instances where we've abstracted the size values into `KemAlgorithm`

Reply:
They don't but the key lengths can be really big (Kyber-512 SK for example is 1632 bytes and Kyber-1024 SK is 3168 bytes). Also aws-lc expects these dynamically allocated

Reply:
I'm not sure if AWS-LC has an expectation for how the memory is allocated. It should be on the user to allocate enough memory either way.
I do agree with the concern for larger key lengths. But I don't have enough experience with how large an array can be before it becomes too "big" to maintain on the stack in this case. It's not a one way door though, we can always look into it later during benchmarking :)